llvm/lib/IR/RepoTicket.cpp

```diff
@@ -165,12 +198,58 @@ std::tuple<bool, unsigned, unsigned> generateTicketMDs(Module &M) {
       return updateDigestGONumAndGetDependencies(GO, GOHash, GOIMap, GONum);
     };
     updateDigestUseCallDependencies(&GO, Hash, Visited, GOIMap, Helper);
+    for (const auto GV : GOIMap[&GO].Contributions) {
+      // Record GO's possible memory dependents.
+      GVIMap[GV].insert(&GO);
+    }
+
     MD5::MD5Result Digest;
     Hash.final(Digest);
     set(&GO, Digest);
     Changed = true;
   }
 
+  if (GVIMap.empty()) {
+    return std::make_tuple(Changed, GONum.VarNum, GONum.FuncNum);
+  }
+
+  // Update the GV's digest using the contributions.
+  for (auto &It : GVIMap) {
+    if (It.first->isDeclaration())
+      continue;
+
+    MD5 GVHash = MD5();
+    GVHash.update(getTicket(It.first)->getDigest().Bytes);
+    // Recursively for all the memory dependent global objects.
+    for (const GlobalObject *GO : It.second) {
+      GVHash.update(getTicket(GO)->getDigest().Bytes);
+    }
+    MD5::MD5Result Digest;
+    GVHash.final(Digest);
+    set(const_cast<GlobalVariable *>(It.first), Digest);
+  }
+
+  // If GO's dependents include the updated GV, update the GO's digest.
+  for (auto &GO : M.global_objects()) {
```

Comment:
I'm slightly bothered by this function. I don't fully understand what it's doing right now, but we've got three significant loops: 

1. Go through the the global objects building `GOIMap` and `GVIMap`, building a hash and setting the MD digest field
2. Iterate over `GVIMap`, computing a hash and set the MD digest field.
3. Similar to the first: it also iterates over the global objects, computing a hash and setting the MD digest field.

Given that there may be a large number of GOs and that setting the MD digest field isn't cheap and that we do it all at least twice, it feels to me like there should be a more optimal implementation. 

Perhaps for a follow-up commit, but I'd like to examine this function more closely.

Reply:
As discussed over the phone, I have created a new task for this issue (https://github.com/SNSystems/llvm-project-prepo/issues/34).